resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnSparkHadoopUtil.scala

```diff
@@ -184,4 +184,29 @@ object YarnSparkHadoopUtil {
     ConverterUtils.toContainerId(containerIdString)
   }
 
+  /**
+   * If MEMORY_OFFHEAP_ENABLED is true, we should ensure executorOverheadMemory requested value
+   * is not less than MEMORY_OFFHEAP_SIZE, otherwise the memory resource requested for executor
+   * may be not enough.
+   */
+  def executorMemoryOverheadRequested(sparkConf: SparkConf): Int = {
+    val executorMemory = sparkConf.get(EXECUTOR_MEMORY).toInt
+    val overhead = sparkConf.get(EXECUTOR_MEMORY_OVERHEAD).getOrElse(
+      math.max((MEMORY_OVERHEAD_FACTOR * executorMemory).toInt, MEMORY_OVERHEAD_MIN)).toInt
+    val offHeap = if (sparkConf.get(MEMORY_OFFHEAP_ENABLED)) {
+      val size =
+        sparkConf.getSizeAsMb(MEMORY_OFFHEAP_SIZE.key, MEMORY_OFFHEAP_SIZE.defaultValueString)
+      require(size > 0,
+        s"${MEMORY_OFFHEAP_SIZE.key} must be > 0 when ${MEMORY_OFFHEAP_ENABLED.key} == true")
+      if (size > overhead) {
+        logWarning(s"The value of ${MEMORY_OFFHEAP_SIZE.key}(${size}MB) will be used as " +
+          s"executorMemoryOverhead to request resource to ensure that Executor has enough memory " +
+          s"to use. It is recommended that the configuration value of " +
+          s"${EXECUTOR_MEMORY_OVERHEAD.key} should be no less than ${MEMORY_OFFHEAP_SIZE.key} " +
+          s"when ${MEMORY_OFFHEAP_ENABLED.key} is true.")
+      }
+      size
+    } else 0
+    math.max(overhead, offHeap).toInt
```

Comment:
I was wondering if it is better to change to `overhead = overhead + offHeap` if off-heap is enabled. Mainly because  off heap memory is not only used for tungsten related, but also for Netty and other native libraries. If we only guarantee `overhead > offHeap`, then it would somehow occupy the usage of Netty and others. Just my two cents :).

Reply:
Got it ~, So Should we add 2 field like `isOffHeapEnabled` and `executorOffHeapMemory` to `YarnAllocator` then use `executorMemory + memoryOverhead + pysparkWorkerMemory + executorOffHeapMemory` to request resource and no longer modify `memoryOverhead`?

Reply:
Hmm, that's a bit complex as for now:

1. If we assume overhead memory includes all the off-heap memory Spark used (include everything). Then user should be aware of the different off-heap memory settings, and carefully set the overhead number to cover all the usages.
2. If we assume that overhead memory only related to some additional memory usage (not explicitly set by Spark, like off-heap memory). Then the overall executor memory should add all as mentioned above.

I think it would be better to involve other's opinion. CC @vanzin @tgravescs .

Reply:
yeah I always thought this was a bit weird off heap was just included in the overhead, but never took the time to go back to see if it was discussed. 

 I think it's better to specifically add the off heap instead of include in the overhead.  Just like we did for the pyspark memory.    executorMemory + memoryOverhead + pysparkWorkerMemory + executorOffHeapMemory. I think that keeps things more consistent and obvious to the user.

Reply:
@tgravescs Agree with you, overhead should  be used to describe memory not use by Spark, like Netty used or JVM used as @jerryshao  said, and we should clearly describe it in the configuration document.

So change to use `executorMemory + memoryOverhead + pysparkWorkerMemory + executorOffHeapMemory` to request resource?

Reply:
In the past,  I would configure `memoryOverhead = executorOffHeapMemory + delta` to ensure memory is enough to use.

Reply:
I'm +1 for changing we can see what others think

Reply:
@tgravescs In our configuration doc I found that `spark.executor.memoryOverhead` described as `Amount of non-heap memory to be allocated per executor process in cluster mode` and `Non-heap memory includes off-heap memory (when spark.memory.offHeap.enabled=true) and memory used by other executor processes (e.g. python process that goes with a PySpark executor) and memory used by other non-executor processes running in the same container.`.

So memoryOverhead = offHeapMemory + pysparkWorkerMemory + otherMemory In the past, now pysparkWorkerMemory has been configured independently, and config offHeapMemory independently is also for consistent.

But there will be some changes in user habits compared with the old version spark.

Reply:
yeah, I understand that, if we are going to change it, 3.0 is a good time to change that behavior.  Like I said, I had found the off heap included in the overhead as confusing because you already had another separate config, why do I as a user have to add it into another config.  Anyway, would be good to get feedback from others as well.
@JoshRosen @beliefer recently had recent changes to docs.

Reply:
@tgravescs Thanks for your ping!

Reply:
@LuciferYang Could you take a look at https://github.com/apache/spark/blob/master/docs/configuration.md

Reply:
@beliefer I know , but now pysparkWorkerMemory already PysparkWorkerMemory is already configured independently use `spark.executor.pyspark.memory` after [21977](https://github.com/apache/spark/pull/21977), now memoryOverhead only contanis  `offHeapMemory ` and `otherMemory`.
 before [c0ef860](https://github.com/apache/spark/pull/25309/commits/c0ef86050b69d34b6f93b569219118e99ec0dd7b) the approach is `if user config offHeapSize > memoryOverhead, use offHeapSize instead of memoryOverhead to request resource and print a warn log to tell user should config memoryOverhead more than offHeapSize`,  there is still no guarantee of request enough memory resource because of the lack of `otherMemory`, and tt's hard to confirm the size of `otherMemory`.

Can we consider `if user config offHeapSize > memoryOverhead, use offHeapSize+384m instead of memoryOverhead to request resource`, `384m is MEMORY_OVERHEAD_MIN`, 
Can we assume that?

Reply:
@beliefer How do we ensure the rationality between `memoryOverhead `  and `offHeapMemory + pysparkWorkerMemory + otherMemory `? Maybe user config as following:
memoryOverhead 2g
offHeapMemory 2g
pysparkWorkerMemory 2g

Reply:
No, `memoryOverhead` contains `spark.executor.pyspark.memory` too. On yarn mode, the total memory of container calculated by process tree. The python process is a sub process.

Reply:
![image](https://user-images.githubusercontent.com/1475305/62510863-677bd880-b843-11e9-9b5f-d5c4098c75db.png)
@beliefer Now `YarnAllocator` line 150 use `executorMemory + memoryOverhead + pysparkWorkerMemory` to new `Resource` Instance, Is this wrong?

Reply:
On the other hand, if the user configures `offheapMemory` and `pysparkWorkerMemory`,
He still needs to configure overhead Memroy and  ensure that the configuration is reasonable(` memoryOverhead > offheapMemory + pysparkWorkerMemory `) in Yarn mode, so that users may need to care about more details.

Reply:
@jerryshao Is the current approach feasible?

Reply:
Let me have a check!

Reply:
I agree with @tgravescs 's opinion.

>yeah, I understand that, if we are going to change it, 3.0 is a good time to change that behavior. Like I said, I had found the off heap included in the overhead as confusing because you already had another separate config, why do I as a user have to add it into another config.

If overhead memory includes off-heap memory, pysparkWorkMemory and others, it makes user hard to set a proper overhead memory, users should know every other settings and figure out a proper number. As of time 3.0, I think we should give a good definition of overhead memory, it can be inconsistent with old version.